// backend/src/scrapers/efectivale.ts
import { decryptString } from '../crypto'
import { chromium, type BrowserContext } from 'playwright'
import { config } from '../config'
import { 
  getSessionMySql, 
  upsertSessionMySql, 
  createManualGasolinaMovement, 
  listTopupRulesMySql,
  listLatestBalances,
  updateWalletBalance
} from '../store/mysql-store'
import path from 'path'
import fs from 'fs'
import { Page } from 'playwright'

/**
 * Extrae el saldo del monedero desde la página actual.
 */
async function extractWalletBalance(page: Page, log: Logger): Promise<number | null> {
  const logLocal: Logger = (m) => log(`[wallet-extract] ${m}`)
  try {
    const text = await page.innerText('.DFC12').catch(() => '')
    if (!text) return null
    const clean = text.replace('$', '').replace(/,/g, '').trim()
    const val = parseFloat(clean)
    if (!isNaN(val)) {
      logLocal(`Saldo Monedero detectado: $${val}`)
      return val
    }
  } catch (e: any) {
    logLocal(`Aviso: No se pudo extraer saldo: ${e.message}`)
  }
  return null
}

// ... (rest of the file until line 287) ...

/**
 * Realiza múltiples recargas en una sola sesión de navegador.
 * Sigue el flujo: Llenar -> Aceptar -> Confirmar -> Verificar -> Siguiente (sin recargar página)
 */
export async function processEfectivaleBulkTopup(args: {
  masterAccount: {
    id: number
    username: string
    passwordEnc: string
    clienteId: string
    consignatarioId: string
  }
  requests: { cuenta: string, monto: number }[]
  log?: Logger
  debug?: boolean
}): Promise<{ results: { cuenta: string, success: boolean, message: string, error?: string }[] }> {
  const log: Logger = (m) => args.log?.(`[bulk-topup] ${m}`)
  const finalResults: any[] = []

  const headless = args.debug || process.env.EFECTIVALE_DEBUG_HEADFUL === 'true' ? false : config.playwrightHeadless
  const browser = await chromium.launch({ headless, slowMo: 100 })
  const ctx = await browser.newContext()

  try {
    const page = await ctx.newPage()
    const password = decryptString(args.masterAccount.passwordEnc)

    // 1. Login Único
    await ensureLoggedIn(page, { ...args.masterAccount, usuario: args.masterAccount.username, password }, log)
    
    // 2. Navegar a Formulario una sola vez
    await gotoDisperisonTarjeta(page, log)

    // 3. Bucle de Dispersión
    for (const req of args.requests) {
      log(`Procesando cuenta: ${req.cuenta} por $${req.monto}...`)
      try {
        // Asegurar que el campo cuenta esté visible (indica que estamos en la pág correcta)
        const accountInput = page.locator('input[name="cuenta"]').first()
        await accountInput.waitFor({ state: 'visible', timeout: 10000 })

        // Llenar datos
        await accountInput.fill(req.cuenta)
        await page.fill('input[name="montoDispersion"]', req.monto.toString())
        
        log(`Enviando Aceptar para ${req.cuenta}`)
        const acceptBtn = page.locator('input[name="xoAceptar"]').first()
        await Promise.all([
          page.waitForLoadState('domcontentloaded'),
          acceptBtn.click(),
        ])
        await page.waitForTimeout(1000)

        // Pantalla de confirmación (Paso 2)
        const confirmText = (await page.innerText('body')).toLowerCase()
        if (confirmText.includes('confirma') || confirmText.includes('aceptar') || confirmText.includes('presione')) {
          log(`Confirmando dispersión para ${req.cuenta}`)
          const confirmBtn = page.locator('input[name="xoAceptar"]').first()
          await Promise.all([
            page.waitForLoadState('domcontentloaded'),
            confirmBtn.click(),
          ])
          // Esperamos a que la franja gris de éxito aparezca
          await page.waitForTimeout(2000)
        }

        // Verificar éxito (Franja gris o texto de éxito)
        const bodyFinal = (await page.innerText('body')).toLowerCase()
        if (
          bodyFinal.includes('éxito') || 
          bodyFinal.includes('exito') || 
          bodyFinal.includes('correctamente') ||
          bodyFinal.includes('realizada')
        ) {
          log(`ÉXITO: ${req.cuenta} dispersada!`)
          finalResults.push({ cuenta: req.cuenta, success: true, message: 'Dispersión completada.' })
          
          // Actualización de Saldo Monedero (Captura del encabezado)
          const currentWallet = await extractWalletBalance(page, log)
          if (currentWallet !== null) await updateWalletBalance(args.masterAccount.id, currentWallet)

          // REGISTRO EN MOVIMIENTOS (Audit Trail)
          try {
             // Traer información completa de la regla y saldos para auditoría
             const allRules = await listTopupRulesMySql();
             const allBalances = await listLatestBalances();
             const rule = allRules.find(r => r.cuenta === req.cuenta);
             const balance = allBalances.find(b => b.cuenta === req.cuenta);

             await createManualGasolinaMovement({
               cuenta: req.cuenta,
               monto: req.monto,
               status: 'dispersed',
               chofer_id: rule?.chofer_id,
               vehiculo_id: rule?.vehiculo_id,
               tarjeta: balance?.tarjeta || rule?.alias, // Fallback al alias si no hay tarjeta
               codigo_card: rule?.short_code?.toString()
             });
          } catch (auditErr) {
             console.error("Error al registrar auditoría (éxito):", auditErr);
          }

        } else {
          log(`ERROR: No se pudo confirmar éxito para ${req.cuenta}`)
          finalResults.push({ cuenta: req.cuenta, success: false, error: 'No se detectó confirmación de éxito.' })
          
          // REGISTRO EN MOVIMIENTOS (Audit Trail - Error)
          try {
             const allRules = await listTopupRulesMySql();
             const allBalances = await listLatestBalances();
             const rule = allRules.find(r => r.cuenta === req.cuenta);
             const balance = allBalances.find(b => b.cuenta === req.cuenta);

             await createManualGasolinaMovement({
               cuenta: req.cuenta,
               monto: req.monto,
               status: 'error',
               error_message: 'No se detectó confirmación de éxito en portal',
               chofer_id: rule?.chofer_id,
               vehiculo_id: rule?.vehiculo_id,
               tarjeta: balance?.tarjeta || rule?.alias,
               codigo_card: rule?.short_code?.toString()
             });
          } catch (auditErr) {
             console.error("Error al registrar auditoría (fallback):", auditErr);
          }
        }

        // PREPARACIÓN PARA SIGUIENTE:
        // En el portal de Efectivale, tras el éxito, a veces hay que dar clic en "Nueva Dispersión"
        // o simplemente el formulario se limpia solo. Según la captura, se queda en la misma página con el formulario listo.
        // Si hay un botón de "Regresar" o similar para limpiar, lo pulsamos.
        const resetBtn = page.locator('input[name="xoCancelar"]').first() // A veces es Cancelar para volver al form limpio
        if (await resetBtn.isVisible().catch(() => false)) {
          // Si vemos que el formulario no está listo, intentamos resetearlo.
          // Pero usualmente tras el mensaje de "Realizada", el portal permite rellenar.
        }

      } catch (e: any) {
        log(`FALLO CRÍTICO en cuenta ${req.cuenta}: ${e.message}`)
        finalResults.push({ cuenta: req.cuenta, success: false, error: e.message })
        
        // Si hubo error de navegación o sesión, intentamos recuperar para la siguiente
        if (page.url().includes('login')) {
          await ensureLoggedIn(page, { ...args.masterAccount, usuario: args.masterAccount.username, password }, log, true)
          await gotoDisperisonTarjeta(page, log)
        }
      }
    }

    return { results: finalResults }

  } finally {
    await browser.close()
  }
}

/**
 * Realiza una recarga única (dispersión) en el portal de Efectivale.
 */
export async function processEfectivaleTopup(args: {
  masterAccount: {
    id: number
    username: string
    passwordEnc: string
    clienteId: string
    consignatarioId: string
  }
  topupRule: {
    cuenta: string
    monto: number
  }
  log?: Logger
  debug?: boolean
}): Promise<{ success: boolean; message: string; error?: string; screenshotPath?: string }> {
  const logs: string[] = []
  const log: Logger = (m) => {
    const line = `[topup-run] ${new Date().toISOString()} ${m}`
    logs.push(line)
    args.log?.(line)
  }

  const headless = args.debug || process.env.EFECTIVALE_DEBUG_HEADFUL === 'true' ? false : config.playwrightHeadless
  const browser = await chromium.launch({ headless, slowMo: 100 })
  const ctx = await browser.newContext()

    let page: any
    try {
      page = await ctx.newPage()
      const password = decryptString(args.masterAccount.passwordEnc)
  
      // 1. Login
      await ensureLoggedIn(page, { ...args.masterAccount, usuario: args.masterAccount.username, password }, log)
    
    // 2. Navegación
    try {
      await gotoDisperisonTarjeta(page, log)
    } catch (e: any) {
      log(`Error navegando a dispersión: ${e.message}`)
      // Si falló navegación, tal vez la sesión expiró a medio camino
      if (page.url().includes('login')) {
         log('Detectada redirección a login durante navegación. Reintentando login...')
         await ensureLoggedIn(page, { ...args.masterAccount, usuario: args.masterAccount.username, password }, log, true)
         await gotoDisperisonTarjeta(page, log)
      } else {
         throw e
      }
    }

    // 3. Llenar Datos Directamente
    log(`Ingresando datos directos - Cuenta: ${args.topupRule.cuenta}, Monto: $${args.topupRule.monto}`)
    
    // Asegurar que estamos en la página correcta y campos visibles
    const accountInput = page.locator('input[name="cuenta"]').first()
    try {
      await accountInput.waitFor({ state: 'visible', timeout: 15000 })
    } catch (e: any) {
      const url = page.url()
      log(`TIMEOUT ERROR: No se encontró "input[name=cuenta]" en ${url}.`)
      // Intentar ver qué hay en la página
      const bodyText = await page.innerText('body').catch(() => 'no body')
      log(`Contenido parcial de la página: ${bodyText.slice(0, 300)}...`)
      throw new Error(`No se encontró el formulario de dispersión en el portal (URL actual: ${url}). Verifique si la sesión expiró o si el portal cambió.`)
    }
    
    // Llenar campos
    await accountInput.fill(args.topupRule.cuenta)
    await page.fill('input[name="montoDispersion"]', args.topupRule.monto.toString())
    
    log('Click Aceptar Inicial (xoAceptar)')
    const acceptBtn = page.locator('input[name="xoAceptar"]').first()
    await Promise.all([
      page.waitForLoadState('domcontentloaded'),
      acceptBtn.click(),
    ])
    await page.waitForTimeout(1500)

    // Verificar si estamos en la pantalla de confirmación
    const confirmText = await page.innerText('body')
    const confirmTextLower = confirmText.toLowerCase()
    
    if (
      confirmTextLower.includes('confirma') || 
      confirmTextLower.includes('aceptar') || 
      confirmTextLower.includes('presione')
    ) {
      log('Pantalla de confirmación detectada (fuzzy match). Segundo Click Aceptar...')
      const confirmBtn = page.locator('input[name="xoAceptar"]').first()
      await Promise.all([
        page.waitForLoadState('domcontentloaded'),
        confirmBtn.click(),
      ])
      await page.waitForTimeout(2500)
    }

    // 5. Verificar resultado final
    const bodyFinal = await page.innerText('body')
    const bodyFinalLower = bodyFinal.toLowerCase()
    
    // Búsqueda más flexible de éxito
    if (
      bodyFinalLower.includes('éxito') || 
      bodyFinalLower.includes('exito') || 
      bodyFinalLower.includes('correctamente') ||
      bodyFinalLower.includes('dispersada') ||
      bodyFinalLower.includes('dispersion realizada')
    ) {
      log('Éxito confirmado (fuzzy match)')
      return { success: true, message: 'Dispersión completada exitosamente.' }
    }

    return { success: false, message: 'No se pudo confirmar el éxito en el portal.', error: bodyFinal.slice(0, 500) }

  } catch (e: any) {
    log(`ERROR: ${e.message}`)
    
    // Capturar screenshot en caso de error
    let screenshotPath: string | undefined
    try {
      const dir = path.join(process.cwd(), 'screenshots')
      if (!fs.existsSync(dir)) fs.mkdirSync(dir)
      screenshotPath = path.join(dir, `error_topup_${Date.now()}.png`)
      await page.screenshot({ path: screenshotPath })
      log(`Screenshot guardado en: ${screenshotPath}`)
    } catch (ssErr) {
      log(`No se pudo tomar screenshot: ${ssErr}`)
    }

    return { success: false, message: 'Fallo en el proceso de dispersión.', error: e.message, screenshotPath }
  } finally {
    await browser.close()
  }
}

export type EfectivaleEmpleadoRow = {
  cuenta: string
  tarjeta: string
  empleado: string
  usuarioParametros: string
  saldo: number | null
}

type StorageState = Awaited<ReturnType<BrowserContext['storageState']>>
type Logger = (msg: string) => void

function parseMoneyToNumber(input: string): number | null {
  const cleaned = input.replace(/\s/g, '').replace(/[$,]/g, '')
  if (!cleaned) return null
  const n = Number(cleaned)
  return Number.isFinite(n) ? n : null
}

async function closeModalIfPresent(page: any) {
  // Cierra modales/banners típicos (bootstrap) y overlays.
  const candidates = [
    'button.close',
    'button[data-dismiss="modal"]',
    'button[aria-label="Close"]',
    'button:has(i.fas.fa-times)',
    'button:has-text("Cerrar")',
    'button:has-text("Aceptar")',
    'button:has-text("Entendido")',
    '.modal-dialog button.close',
    '.modal-dialog button[aria-label="Close"]',
  ]

  for (let attempt = 0; attempt < 5; attempt++) {
    let closedAny = false

    // Caso específico observado: #myModal intercepta clicks.
    const myModal = page.locator('#myModal').first()
    if (await myModal.isVisible().catch(() => false)) {
      // Forzar a que el modal no intercepte clicks mientras lo cerramos.
      await page
        .evaluate(() => {
          const d: any = (globalThis as any).document
          const m = d?.querySelector?.('#myModal') as any
          if (m) m.style.pointerEvents = 'none'
          d?.querySelectorAll?.('.modal-backdrop')?.forEach?.((b: any) => (b.style.pointerEvents = 'none'))
        })
        .catch(() => {})

      // Intentar cerrar desde el modal
      const closeInside = myModal.locator(candidates.join(',')).first()
      if (await closeInside.isVisible().catch(() => false)) {
        await closeInside.click({ timeout: 2000 }).catch(() => {})
        closedAny = true
        await page.waitForTimeout(250)
      } else {
        // Intentar ESC
        await page.keyboard.press('Escape').catch(() => {})
        await page.waitForTimeout(250)
      }

      // Si aún visible, forzar ocultamiento del modal/backdrop
      if (await myModal.isVisible().catch(() => false)) {
        await page
          .evaluate(() => {
            const d: any = (globalThis as any).document
            if (!d) return
            const m = d.querySelector('#myModal') as any
            if (m) {
              m.classList.remove('in')
              m.style.display = 'none'
              m.setAttribute('aria-hidden', 'true')
            }
            d.querySelectorAll('.modal-backdrop').forEach((b: any) => b.remove())
            d.body?.classList?.remove('modal-open')
          })
          .catch(() => {})
        closedAny = true
        await page.waitForTimeout(250)
      }
    }

    for (const sel of candidates) {
      const btn = page.locator(sel).first()
      if (await btn.isVisible().catch(() => false)) {
        await btn.click({ timeout: 2000 }).catch(() => {})
        closedAny = true
        await page.waitForTimeout(250)
      }
    }
    if (!closedAny) break
  }
}

async function ensureLoggedIn(
  page: any,
  creds: { clienteId: string; consignatarioId: string; usuario: string; password: string },
  log: Logger,
  force: boolean = false
) {
  log('Verificando estado de sesión...')
  const currentUrl = page.url()
  const isLoginPage = currentUrl.includes('login')
  const loginFormVisible = await page.locator('input[name="clienteID"]').isVisible().catch(() => false)
  
  if (!force && !isLoginPage && !loginFormVisible) {
    // Si no estamos en login y el formulario no es visible, probamos si el menú existe
    const hasMenu = await page.getByText('Menú Efectinet').isVisible().catch(() => false)
    if (hasMenu) {
      log('Sesión parece activa (Menú detectable)')
      return
    }
  }

  log('Iniciando proceso de Login...')
  await page.goto('https://www.efectivale.com.mx/efectinet/login', { waitUntil: 'domcontentloaded' })
  await page.waitForTimeout(1000)

  log('Llenar campos de login')
  await page.fill('input[name="clienteID"]', creds.clienteId)
  await page.fill('input[name="consignatarioID"]', creds.consignatarioId)
  await page.fill('input[name="usuarioUSR"]', creds.usuario)
  await page.fill('input[name="usuarioPWD"]', creds.password)

  log('Click Ingresar')
  await Promise.all([
    page.waitForNavigation({ waitUntil: 'networkidle', timeout: 30000 }).catch(() => {}),
    page.click('input[name="Ingresar"]'),
  ])

  log('Cerrar modal si aparece')
  await closeModalIfPresent(page)
  
  // Verificación final de login
  const loggedIn = await page.getByText('Menú Efectinet').isVisible().catch(() => false)
  if (!loggedIn) {
    log('ADVERTENCIA: No se detectó el menú principal tras login.')
  } else {
    log('Login exitoso confirmado.')
  }
}

async function gotoDisperisonTarjeta(page: any, log: Logger) {
  // Menú Efectinet -> Administración de Servicios -> Efecticard Corporativo ->
  // Administración Monedero Corporativo -> Dispersión Tarjeta Corporativo
  // La UI es por divs con onclick; usar texto es más robusto.
  log('Abrir menú: Menú Efectinet')
  await closeModalIfPresent(page)
  // Aumentar timeout y reintentar si es necesario
  await page.getByText('Menú Efectinet', { exact: true }).click({ timeout: 60000, force: true }).catch(async (e: any) => {
    log('Timeout en menú principal, intentando recargar página...')
    await page.reload()
    await page.getByText('Menú Efectinet', { exact: true }).click({ timeout: 30000, force: true })
  })
  log('Click: Administración de Servicios')
  await closeModalIfPresent(page)
  await page.getByText('Administración de Servicios', { exact: true }).click({ timeout: 20000, force: true })
  log('Click: Efecticard Corporativo')
  await closeModalIfPresent(page)
  await page.getByText('Efecticard Corporativo', { exact: true }).click({ timeout: 20000, force: true })
  log('Click: Administración Monedero Corporativo')
  await closeModalIfPresent(page)
  await page.getByText('Administración Monedero Corporativo', { exact: true }).click({ timeout: 20000, force: true })
  log('Click: Dispersión Tarjeta Corporativo')
  await closeModalIfPresent(page)
  await page.getByText('Dispersión Tarjeta Corporativo', { exact: true }).click({ timeout: 20000, force: true })
  await page.waitForTimeout(700)
}

function extractEmpleadosFromTable(htmlTexts: string[][]): EfectivaleEmpleadoRow[] {
  const out: EfectivaleEmpleadoRow[] = []
  for (const tds of htmlTexts) {
    // layout: [radio, Cuenta, Tarjeta, Empleado, Usuario Parametros, Saldo]
    if (tds.length < 6) continue
    const cuenta = tds[1]?.trim() ?? ''
    const tarjeta = tds[2]?.trim() ?? ''
    const empleado = tds[3]?.trim() ?? ''
    const usuarioParametros = tds[4]?.trim() ?? ''
    const saldo = parseMoneyToNumber(tds[5] ?? '')
    if (cuenta) out.push({ cuenta, tarjeta, empleado, usuarioParametros, saldo })
  }
  return out
}

async function readCurrentPageRows(page: any): Promise<EfectivaleEmpleadoRow[]> {
  // Usar textContent para no perder valores cuando hay spans/nbsp.
  const texts: string[][] = await page.$$eval('table.DT tr.DF1, table.DT tr.DF2', (trs: any[]) =>
    trs.map((tr) =>
      Array.from(tr.querySelectorAll('td')).map((td) =>
        (String((td as any).textContent ?? '')).replace(/\u00a0/g, ' ').replace(/\s+/g, ' ').trim()
      )
    )
  )

  // Algunas variantes agregan columnas; usamos el último TD como saldo.
  const normalized = texts.map((tds) => {
    if (tds.length < 6) return tds
    const saldo = tds[tds.length - 1]
    const base = tds.slice(0, 5)
    return [...base, saldo]
  })

  return extractEmpleadosFromTable(normalized)
}

async function getPagerText(page: any): Promise<string> {
  const pager = page.locator('td.IFT').first()
  return (await pager.innerText().catch(() => '')).trim()
}

export async function scrapeEfectivaleEmpleados(args: {
  creds: { clienteId: string; consignatarioId: string; usuario: string; password: string }
  storageState?: StorageState
  log?: Logger
  debug?: boolean
}): Promise<{ rows: EfectivaleEmpleadoRow[]; storageState: StorageState; logs: string[] }> {
  const logs: string[] = []
  const log: Logger = (m) => {
    const line = `[efectivale] ${new Date().toISOString()} ${m}`
    logs.push(line)
    args.log?.(line)
  }

  const headless = args.debug || process.env.EFECTIVALE_DEBUG_HEADFUL === 'true' ? false : config.playwrightHeadless
  const slowMo = Number(process.env.EFECTIVALE_SLOWMO_MS ?? 0) || 0
  const browser = await chromium.launch({ headless, slowMo })
  const ctx = await browser.newContext(args.storageState ? { storageState: args.storageState } : {})
  try {
    const page = await ctx.newPage()
    let step = 'login'
    try {
      await ensureLoggedIn(page, args.creds, log)
      step = 'menu'
      await gotoDisperisonTarjeta(page, log)
      step = 'table'
    } catch (e: any) {
      throw new Error(`Fallo en paso "${step}": ${e?.message ?? String(e)}`)
    }

    const all: EfectivaleEmpleadoRow[] = []
    const seen = new Set<string>()

    let lastPager = ''
    for (let guard = 0; guard < 60; guard++) {
      log(`Leer página ${guard + 1}`)
      const pager = await getPagerText(page)
      if (pager && pager === lastPager) break
      lastPager = pager

      const rows = await readCurrentPageRows(page)
      for (const r of rows) {
        const key = `${r.cuenta}-${r.tarjeta}`
        if (!seen.has(key)) {
          seen.add(key)
          all.push(r)
        }
      }

      const nextBtn = page.locator('input[name="xoSiguiente"]').first()
      const canNext = await nextBtn.isVisible().catch(() => false)
      if (!canNext) break

      log('Click: Siguiente')
      await Promise.all([
        page.waitForLoadState('domcontentloaded'),
        nextBtn.click(),
      ])
      await page.waitForTimeout(400)
    }

    const storageState = await ctx.storageState()
    log(`OK. Filas totales: ${all.length}`)
    return { rows: all, storageState, logs }
  } finally {
    await browser.close()
  }
}

/**
 * Orquestador que maneja sesiones en MySQL y scrapea empleados/saldos de una cuenta.
 */
export async function scrapeEfectivaleSaldoSnapshotAsJson(args: {
  accountId: number
  creds: { clienteId: string; consignatarioId: string; usuario: string; password: string }
  log?: (m: string) => void
  debug?: boolean
}) {
  // 1. Obtener sesión persistente si existe
  const storageState = await getSessionMySql(args.accountId)
  
  // 2. Ejecutar scrap
  let playwrightStorageState: any = undefined
  if (storageState?.storage_state_json) {
    try {
      playwrightStorageState = JSON.parse(storageState.storage_state_json)
    } catch (e) {
      args.log?.(`Error parseando sesión previa: ${String(e)}`)
    }
  }

  const result = await scrapeEfectivaleEmpleados({
    creds: args.creds,
    storageState: playwrightStorageState,
    log: args.log,
    debug: args.debug
  })

  // 3. Persistir nueva sesión
  if (result.storageState) {
    await upsertSessionMySql(args.accountId, JSON.stringify(result.storageState))
  }

  // 4. Retornar filas
  return result.rows
}

/**
 * Consulta solo el saldo del monedero de una cuenta maestra.
 */
export async function fetchEfectivaleWalletBalance(args: {
  masterAccount: {
    id: number
    username: string
    passwordEnc: string
    clienteId: string
    consignatarioId: string
  }
  log?: Logger
  debug?: boolean
}): Promise<number | null> {
  const log: Logger = (m) => args.log?.(`[wallet-fetch] ${m}`)
  const headless = args.debug || process.env.EFECTIVALE_DEBUG_HEADFUL === 'true' ? false : config.playwrightHeadless
  const browser = await chromium.launch({ headless })
  try {
    const ctx = await browser.newContext()
    const page = await ctx.newPage()
    const password = decryptString(args.masterAccount.passwordEnc)

    await ensureLoggedIn(page, { ...args.masterAccount, usuario: args.masterAccount.username, password }, log)
    await gotoDisperisonTarjeta(page, log)
    
    const balance = await extractWalletBalance(page, log)
    if (balance !== null) {
      await updateWalletBalance(args.masterAccount.id, balance)
    }
    return balance
  } finally {
    await browser.close()
  }
}
