// backend/src/modules/admin.ts
import { Router } from 'express'
import {
  getGasolinaRequest,
  listGasolinaRequests,
  mysqlEnabled,
  updateGasolinaRequest,
  listTopupRulesMySql,
  upsertTopupRuleMySql,
  deleteTopupRuleMySql,
  seedRulesFromSaldos,
  listLatestBalances,
  listChoferes,
  getChoferHistory,
  createChofer,
  updateChofer,
  deleteChofer,
  assignAssetsToChofer,
  listVehiculos,
  createVehiculo,
  updateVehiculo,
  deleteVehiculo,
  listAccountsMySql,
} from '../store/mysql-store'
import { runSingleTopupRequest } from '../services/telegram-helpers'
import { notifyTelegramUser } from '../services/telegram-notify'
import { 
  scrapeEfectivaleSaldoSnapshotAsJson, 
  processEfectivaleTopup,
  processEfectivaleBulkTopup,
  fetchEfectivaleWalletBalance 
} from '../scrapers/efectivale'

export const adminRouter = Router()

// --- CUENTAS (ACCESOS) ---
adminRouter.get('/accounts', async (_req, res) => {
  try {
    const rows = await listAccountsMySql()
    res.json(rows)
  } catch (e: any) {
    res.status(500).json({ error: e?.message ?? String(e) })
  }
})

adminRouter.get('/wallet-balance/:accountId', async (req, res) => {
  const accountId = Number(req.params.accountId)
  const accounts = await listAccountsMySql()
  const masterAcc = accounts.find(a => a.id === accountId)
  
  if (!masterAcc) return res.status(404).json({ error: 'Cuenta maestra no encontrada' })

  const { clienteId, consignatarioId } = JSON.parse(masterAcc.extra_json ?? '{}')

  try {
     const balance = await fetchEfectivaleWalletBalance({
       masterAccount: {
         id: masterAcc.id,
         username: masterAcc.username,
         passwordEnc: masterAcc.password_enc,
         clienteId,
         consignatarioId,
       },
       debug: process.env.EFECTIVALE_DEBUG_HEADFUL === 'true'
     })
     res.json({ ok: true, balance })
  } catch (e: any) {
    res.status(500).json({ error: e.message })
  }
})

// --- CHOFERES ---
adminRouter.get('/choferes', async (_req, res) => {
  try {
    const rows = await listChoferes()
    res.json(rows)
  } catch (e: any) {
    res.status(500).json({ error: e?.message ?? String(e) })
  }
})

adminRouter.post('/choferes', async (req, res) => {
  try {
    const id = await createChofer(req.body.nombre)
    res.json({ ok: true, id })
  } catch (e: any) {
    res.status(500).json({ error: e?.message ?? String(e) })
  }
})

adminRouter.put('/choferes/:id', async (req, res) => {
  try {
    await updateChofer(Number(req.params.id), req.body.nombre)
    res.json({ ok: true })
  } catch (e: any) {
    res.status(500).json({ error: e?.message ?? String(e) })
  }
})

adminRouter.delete('/choferes/:id', async (req, res) => {
  try {
    await deleteChofer(Number(req.params.id))
    res.json({ ok: true })
  } catch (e: any) {
    res.status(500).json({ error: e?.message ?? String(e) })
  }
})

adminRouter.put('/choferes/:id/assets', async (req, res) => {
  try {
    const { vehiculo_id, cuenta } = req.body
    await assignAssetsToChofer(Number(req.params.id), vehiculo_id || null, cuenta || null)
    res.json({ ok: true })
  } catch (e: any) {
    res.status(500).json({ error: e?.message ?? String(e) })
  }
})

adminRouter.get('/choferes/:id/history', async (req, res) => {
  try {
    const history = await getChoferHistory(Number(req.params.id))
    res.json(history)
  } catch (e: any) {
    res.status(500).json({ error: e?.message ?? String(e) })
  }
})

// --- VEHÍCULOS ---
adminRouter.get('/vehiculos', async (_req, res) => {/*  */
  try {
    const rows = await listVehiculos()
    res.json(rows)
  } catch (e: any) {
    res.status(500).json({ error: e?.message ?? String(e) })
  }
})

adminRouter.post('/vehiculos', async (req, res) => {
  try {
    const id = await createVehiculo(req.body)
    res.json({ ok: true, id })
  } catch (e: any) {
    res.status(500).json({ error: e?.message ?? String(e) })
  }
})

adminRouter.put('/vehiculos/:id', async (req, res) => {
  try {
    await updateVehiculo(Number(req.params.id), req.body)
    res.json({ ok: true })
  } catch (e: any) {
    res.status(500).json({ error: e?.message ?? String(e) })
  }
})

adminRouter.delete('/vehiculos/:id', async (req, res) => {
  try {
    await deleteVehiculo(Number(req.params.id))
    res.json({ ok: true })
  } catch (e: any) {
    res.status(500).json({ error: e?.message ?? String(e) })
  }
})

function adminAuth(req: any, res: any, next: any) {
  const expected = process.env.ADMIN_API_KEY
  // Si no hay clave configurada en el servidor, permitimos acceso libre (uso local)
  if (!expected) {
    return next()
  }
  const key = String(req.headers['x-admin-key'] ?? '')
  if (key !== expected) {
    res.status(401).json({ error: 'No autorizado. Se requiere x-admin-key válida.' })
    return
  }
  next()
}

adminRouter.use(adminAuth)

adminRouter.get('/mysql-status', (_req, res) => {
  res.json({ enabled: mysqlEnabled() })
})

adminRouter.get('/balances', async (_req, res) => {
  try {
    const rows = await listLatestBalances()
    res.json(rows)
  } catch (e: any) {
    res.status(500).json({ error: e?.message ?? String(e) })
  }
})

adminRouter.get('/gasolina', async (req, res) => {
  try {
    const status = typeof req.query.status === 'string' ? req.query.status : undefined
    const rows = await listGasolinaRequests(status)
    res.json(rows)
  } catch (e: any) {
    res.status(500).json({ error: e?.message ?? String(e) })
  }
})

adminRouter.post('/gasolina/:id/approve', async (req, res) => {
  try {
    const row = await getGasolinaRequest(req.params.id)
    if (!row || row.status !== 'pending') {
      res.status(404).json({ error: 'Solicitud no encontrada o ya procesada' })
      return
    }
    const note = req.body?.note != null ? String(req.body.note) : null
    const payload: {
      cuenta?: string
      tarjetaUltimos7?: string
      requestedMonto: number
      debug?: boolean
    } = {
      requestedMonto: row.monto,
      debug: false,
    }
    if (row.id_tipo === 'cuenta') {
      if (!row.cuenta) {
        res.status(400).json({ error: 'Solicitud sin cuenta' })
        return
      }
      payload.cuenta = row.cuenta
    } else {
      if (!row.tarjeta_ultimos7) {
        res.status(400).json({ error: 'Solicitud sin tarjeta' })
        return
      }
      payload.tarjetaUltimos7 = row.tarjeta_ultimos7
    }

    const result = await runSingleTopupRequest(payload)
    await updateGasolinaRequest(row.id, { 
      status: 'dispersed', 
      admin_note: note, 
      error_message: null,
      admin_approver_name: 'Dashboard Admin',
      admin_approver_id: 1
    })
    await notifyTelegramUser(
      row.telegram_chat_id,
      `Tu carga de gasolina ya fue dispersada.\n${result.message}\n\nGracias por usar mis servicios.`
    )
    res.json({ ok: true, result })
  } catch (e: any) {
    try {
      const row = await getGasolinaRequest(req.params.id)
      if (row) {
        await updateGasolinaRequest(row.id, { 
          status: 'error', 
          error_message: e?.message ?? String(e),
          admin_approver_name: 'Dashboard Admin (Failed)',
          admin_approver_id: 1
        })
      }
      if (row) {
        await notifyTelegramUser(
          row.telegram_chat_id,
          `No se pudo completar tu recarga: ${e?.message ?? e}\n\nGracias por usar mis servicios.`
        )
      }
    } catch {
      // ignore
    }
    res.status(502).json({ error: e?.message ?? String(e) })
  }
})

adminRouter.post('/gasolina/:id/reject', async (req, res) => {
  try {
    const row = await getGasolinaRequest(req.params.id)
    if (!row || row.status !== 'pending') {
      res.status(404).json({ error: 'Solicitud no encontrada o ya procesada' })
      return
    }
    const note = req.body?.note != null ? String(req.body.note) : null
    await updateGasolinaRequest(row.id, { 
      status: 'rejected', 
      admin_note: note,
      admin_approver_name: 'Dashboard Admin',
      admin_approver_id: 1
    })
    await notifyTelegramUser(
      row.telegram_chat_id,
      `Tu solicitud de gasolina no fue autorizada.${note ? ` Nota: ${note}` : ''}\n\nGracias por usar mis servicios.`
    )
    res.json({ ok: true })
  } catch (e: any) {
    res.status(500).json({ error: e?.message ?? String(e) })
  }
})

adminRouter.post('/disperse-bulk', async (req, res) => {
  const requests: { cuenta: string, monto: number, accountId?: number }[] = req.body.requests || []
  if (requests.length === 0) return res.json({ ok: true, results: [] })

  const accounts = await listAccountsMySql()
  const results: any[] = []

  // Agrupar por accountId para procesar en bloques (un navegador por cuenta maestra)
  const groups: Record<number, typeof requests> = {}
  for (const r of requests) {
    const accId = r.accountId || accounts[0]?.id
    if (!accId) continue
    if (!groups[accId]) groups[accId] = []
    groups[accId].push(r)
  }

  for (const accIdStr of Object.keys(groups)) {
    const accId = Number(accIdStr)
    const accRequests = groups[accId]
    const masterAcc = accounts.find(a => a.id === accId)

    if (!masterAcc) {
      accRequests.forEach(r => results.push({ cuenta: r.cuenta, success: false, error: 'Cuenta maestra no encontrada' }))
      continue
    }

    const { clienteId, consignatarioId } = JSON.parse(masterAcc.extra_json ?? '{}')
    
    try {
      const bulkResult = await processEfectivaleBulkTopup({
        masterAccount: {
          id: masterAcc.id,
          username: masterAcc.username,
          passwordEnc: masterAcc.password_enc,
          clienteId,
          consignatarioId,
        },
        requests: accRequests.map(r => ({ cuenta: r.cuenta, monto: r.monto })),
        debug: process.env.EFECTIVALE_DEBUG_HEADFUL === 'true'
      })
      results.push(...bulkResult.results)
    } catch (e: any) {
      accRequests.forEach(r => results.push({ cuenta: r.cuenta, success: false, error: e.message }))
    }
  }

  // AL FINALIZAR: Disparamos un sync rápido para refrescar los saldos
  try {
     // No esperamos a que termine para no colgar la respuesta, se hace en "background"
     // o si prefieres confirmación, podrías esperar. Por UX, lanzamos el background.
     const topAccounts = Object.keys(groups).map(Number)
     for(const accId of topAccounts) {
        const masterAcc = accounts.find(a => a.id === accId)
        if (masterAcc) {
           const creds = JSON.parse(masterAcc.extra_json || '{}')
           creds.usuario = masterAcc.username
           creds.password = masterAcc.password_enc // decrypted in scraper
           scrapeEfectivaleSaldoSnapshotAsJson({
             accountId: accId,
             creds: { ...creds },
             debug: false
           }).catch(e => console.error("Error en sync automático post-carga:", e))
        }
     }
  } catch (e) {
    console.error("No se pudo iniciar sync automático:", e)
  }
  
  res.json({ ok: true, results })
})

// --- REGLAS DE RECARGA ---

adminRouter.get('/rules', async (_req, res) => {
  try {
    const rows = await listTopupRulesMySql()
    res.json(rows)
  } catch (e: any) {
    res.status(500).json({ error: e?.message ?? String(e) })
  }
})

adminRouter.post('/rules', async (req, res) => {
  try {
    const id = await upsertTopupRuleMySql(req.body)
    res.json({ ok: true, id })
  } catch (e: any) {
    res.status(500).json({ error: e?.message ?? String(e) })
  }
})

adminRouter.delete('/rules/:id', async (req, res) => {
  try {
    await deleteTopupRuleMySql(Number(req.params.id))
    res.json({ ok: true })
  } catch (e: any) {
    res.status(500).json({ error: e?.message ?? String(e) })
  }
})

adminRouter.post('/rules/seed', async (_req, res) => {
  try {
    const count = await seedRulesFromSaldos()
    res.json({ ok: true, count })
  } catch (e: any) {
    res.status(500).json({ error: e?.message ?? String(e) })
  }
})
